test/performance/specs/site-editor.spec.js

```diff
@@ -193,7 +193,7 @@ test.describe( 'Site Editor Performance', () => {
 				await metrics.startTracing();
 
 				await page
-					.getByRole( 'button', { name: 'Single Posts' } )
+					.getByRole( 'link', { name: 'Single Posts' } )
```

Comment:
I think this is probably the change that is breaking the performance tests in this PR. 
If that change is required, we need to keep both "button" and "link" because performance tests compare two branches and old ones would have a "button" while this one will require "link". So we need something that seems button or link here.

Reply:
Oh, thanks!

What is this test tracking? The time it takes an editor to load since an item is clicked?

Could I use a different path (loading pages or parts instead of templates)? If I understand this correctly, I see a couple of potential issues with that approach:

1. it will change the loading profile for the test, so it can render different results just because we are loading different things than before
2. it'll work _for this PR_ because it compares with `trunk`, however the issue will happen again when it is compared against the _base commit_ (during the plugin release, I believe).

So, my options boil down to use some kind of flag/check to run one code path or the other in the test (I'm not sure that's even possible?) or find a common locator that works for these two screens:

| Trunk | This PR |
| --- | --- |
| <img width="1508" alt="Captura de ecrã 2024-03-15, às 10 41 01" src="https://github.com/WordPress/gutenberg/assets/583546/41892fe4-41bc-4e52-befa-aa667e9977aa"> | <img width="1508" alt="Captura de ecrã 2024-03-15, às 10 38 52" src="https://github.com/WordPress/gutenberg/assets/583546/afadedaa-862b-4c0c-8473-b19fb0a3ea57"> |

Is this correct?

Going to start looking for the locator. I'd welcome looking at other cases where we ran into the same issue to get inspiration from.

Reply:
[getByText](https://playwright.dev/docs/locators#locate-by-text) works locally, let's what CI thinks of it.

Reply:
> it will change the loading profile for the test, so it can render different results just because we are loading different things than before

Yes, it compares the time it takes to switch between two templates while the preview is visible. I realize that this is changing. So for legacy "template screens" it should continue to work as is, for new ones it's probably better to check in the "list" data view right?

I mean we can change it to pages or others if needed but it wouldn't solve the problem, it will just delay solving it. (when the dataviews get lifted up for these)

Reply:
So in other words, we should add a check to the test and if we're in an old branch keep using the regular root /template but if we're in a new branch, we should switch to list view first before navigating. Otherwise comparing old vs new wouldn't make sense.

Reply:
👍 I may be able to check for the actions menu visibility as a proxy: if it's there, switch to list view; otherwise, use the old test.

Reply:
https://github.com/WordPress/gutenberg/pull/59858 would be benefitial to land first, as it makes the list layout available without the experiment enabled

Reply:
https://github.com/WordPress/gutenberg/pull/59792/commits/0d7b67e3dc0d377ecd16b635089987b844f199b0

Reply:
Looks good, I'm curious what the results will yield here and if we'll see a jump in performance.

Reply:
Navigate is 5x slower, according to the test. 🤔 
<img width="615" alt="Captura de ecrã 2024-03-15, às 13 20 51" src="https://github.com/WordPress/gutenberg/assets/583546/79e5fab8-19c0-47d0-b03d-6509a0c3c20b">

Reply:
I've reviewed how the new test works (`npm run test:performance:debug -- site-editor -g "Navigating"`) and it seems to do so as expected.

I'll need to look more into how metric utils for e2e test work to understand what we were measuring before and after — and if we are actually measuring the same thing.

Looking into devtools > performance for how this branch works vs trunk I don't see anything obvious (attach the traces for this branch and trunk, [devtools-performance-traces.zip](https://github.com/WordPress/gutenberg/files/14615635/devtools-performance-traces.zip)). The click event triggers a ~475ms execution in both this branch and trunk (though with some differences):

| Branch | Trunk |
| --- | --- |
| <img width="1336" alt="Captura de ecrã 2024-03-15, às 13 51 56" src="https://github.com/WordPress/gutenberg/assets/583546/dee792d4-295b-41ef-8c8b-1b73a5929943"> | <img width="1329" alt="Captura de ecrã 2024-03-15, às 13 51 28" src="https://github.com/WordPress/gutenberg/assets/583546/842f0fc8-b794-4ca3-ba84-1a3b5c2fc2ed"> |

My suggestion is merging this PR and investigate this independently.

Reply:
The screenshot you shared is actually very telling :)

You're right that the overall time is similar but the "click" event which is what we're measuring is way bigger in this PR. In other words, something that was async (like in a micro task or delayed using a promise or something) is becoming sync in this PR.

I don't necessarily think that's a bad thing on its own but it shows that we need to update our performance test to compute the whole stack (clicking and re-rendering) regardless of whether the rendering is async (for some reason).

So I agree that we should merge and try to improve the metric separately.

Reply:
So, I've looked at how/what we measure. [getClickEventDurations](https://github.com/WordPress/gutenberg/blob/trunk/test/performance/specs/site-editor.spec.js#L203) ends up calling [getEventDurations](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-test-utils-playwright/src/metrics/index.ts#L275) filtering by `click` event type. If I understand how this is implemented, it means we were/are measuring the _duration_ of the `click` event and not the whole execution (other tasks). In the traces I've shared, the `click` event is 210ms in this branch vs 57ms in trunk: a ~x4 performance impact, more in line with the numbers reported by our test.